internal/native_image/common.bzl

```diff
@@ -162,6 +172,13 @@ def _prepare_native_image_rule_context(
     else:
         path_list_separator = ":"
 
+    # if we are building a shared library, headers will be emitted; by default, the `graal_isolate.h`
+    # and `graal_isolate_dynamic.h` files are included. additional headers can be added by the `out_headers`
+    # attribute.
+    if ctx.attr.shared_library and ctx.attr.default_outputs:
+        additional_outputs.append(_GRAALVM_ISOLATE_HEADER)
```

Comment:
these headers are produced automatically by the `native-image` build -- so if they aren't added here, they end up as undeclared outputs

Reply:
Yeah, I'm in favor of adding them here. But maybe we could just get rid of `default_outputs` and have this behavior in place unconditionally?

Reply:
I guess I'm concerned about multiple shared library targets within the same package, but that would be something the user has to deal with anyway?

In my downstream test codebase, I ended up needing to manage `graal_*.h` headers manually, because _that_ codebase expected angle syntax for importing headers. So I needed to pull them into their own `cc_library` target with an `include =` path set.

If the headers are exported unconditionally by the rule, I worry there would be no easy way to prevent a file collision under this use case. Maybe I'm missing something.

The _default_ functionality for the rules seems to align with including them as defaults, otherwise a user would have to declare them unconditionally, so I agree this isn't ideal and would love a better way.

FWIW I think I may merge `additional_outputs` and `out_headers`, but those may need to remain separate so that we can attach headers to providers. We could also sniff the filename, if that's preferable let me know. Open to your advice on all aspects above.

Reply:
Could we separate the headers by having them emitted into a subdirectory? E.g., if we declare the headers for target `foo` with package relative path `foo/graal_something.h`, we can have multiple native images coexist in a package and also provide a way to disambiguate headers via the prefix.

Reply:
I think I like separate attributes better than file path sniffing

Reply:
I can check for hidden/expert `native-image` flags, but AFAIK those are calculated from the binary name, so they are already under the user's control via `executable_name`. I'm not sure about a subfolder with `executable_name`, I think we'd have to slice that and append it to `-H:OutputPath`, because the binary name is passed separately from the output prefix (to `native-image`).

In any case, I agree with you, I'd like to find a clener way to do it.

Reply:
(Lots of this is telling me that a dedicated `shared_native_image` rule might be a good idea, by the way, because the shared library function of `native-image` demands radically different flags than an executable, as evidenced by `executable_name` being a little wonky in this case. Just food for later thought, your ideas are always welcome of course.)

Reply:
Yes, I also like a dedicated rule better. For example, such a rule could be marked as non-executable and could also statically declare that it emits `CcSharedLibraryInfo`.

Reply:
Sorry, to clarify, the regular header names are under the user's control, and follow the pattern of the `executable_name` (an executable name of `libsample` gets `libsample.so` and `libsample.h`, etc).

The standard GraalVM headers, i.e. `graal_isolate.h` and family, are generated unconditionally by `native-image` at those names. I don't see a way to override those.

Reply:
Could we prepend `foo/` to `shared_library_name` and then have a separate action symlink it to the location without the prefix? That would get us prefixed headers without the user having to refer to the shared library with the prefix.